src/models.py

```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
         self.device = device
         
         self.emb = nn.Embedding(vocab_size, embedding_dim)
-        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional)
-        self.fc = nn.Linear(num_hidden, vocab_size)
+        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional) 
+        self.attn = nn.Linear(self.num_hidden * 2, 1) 
+        self.fc = nn.Linear(self.num_hidden * 2, vocab_size)
     
-    def forward(self, x, h0, seq_lengths):
-        # x has shape (batch, seq_len), h0 has shape (num_layers * num_directions, batch, num_hidden)
+    def forward(self, caption, h0, c0, seq_lengths, intermediate_layer):
+        # -------------- Inputs --------------------
+            # caption = caption; shape = (batch, seq_len)
+            # mlp = image feature vector averaged and passed through MLP
+            # seq_lengths = ''
+            # intermediate_layer = output of 'encoder.layer4[2].bn2' 
         
-        c0 = self.init_cell_state(h0.shape[1])
-        
-        emb = self.emb(x)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+        # Initialising the hidden state and cell state of the LSTM with 
+        # the two different MLPs
+
+        c0, h0 = self.init_cell_state_and_hidden_state(caption.shape[0], c0, h0)
         
+        emb = self.emb(caption)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+     
         # Pack_padded_sequence so that padded items in the sequence won't be shown to the LSTM
         # See https://towardsdatascience.com/taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
         # We use seq_lengths - 1 because we do not want to count the last <END> token in each caption.
         emb = torch.nn.utils.rnn.pack_padded_sequence(emb, seq_lengths - 1, batch_first=True)
+
+        # [2, 38, 512]
+
+        # input of shape (seq_len, batch, input_size):
         
         out, _ = self.lstm(emb, (h0, c0))
 
         # Undo the packing operation
         out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # (batch, seq_len, num_directions * num_hidden)
         
```

Comment:
Since you reuse the shape parameters of the `out` and `intermediate_layer` tensor several times, I would assign them to variables. This also helps readability, since it's then easier to see what the shapes represent. For example,  you can add 

```suggestion

batch, seq_len, directions_times_hidden = out.shape
_, _, n_annotations = intermediate_layer.shape
```

Then  you can also use these variables from now on, because for me it can be hard to read when you constantly  refer back to the shape parameters of for example `out` or `intermediate_layer`, which I then have to look up.